File: library/src/main/java/com/raizlabs/android/dbflow/sql/language/Set.java
package com.raizlabs.android.dbflow.sql.language;

import android.content.ContentValues;
import android.database.Cursor;

import com.raizlabs.android.dbflow.config.FlowManager;
import com.raizlabs.android.dbflow.sql.Queriable;
import com.raizlabs.android.dbflow.sql.Query;
import com.raizlabs.android.dbflow.sql.QueryBuilder;
import com.raizlabs.android.dbflow.sql.builder.Condition;
import com.raizlabs.android.dbflow.sql.builder.ConditionQueryBuilder;
import com.raizlabs.android.dbflow.structure.Model;

import java.util.Map;

/**
 * Description: Used to specify the SET part of an {@link com.raizlabs.android.dbflow.sql.language.Update} query.
 */
public class Set<ModelClass extends Model> implements WhereBase<ModelClass>, Queriable {

    private ConditionQueryBuilder<ModelClass> mConditionQueryBuilder;

    private Query mUpdate;

    Set(Query update, Class<ModelClass> table) {
        mUpdate = update;
        mConditionQueryBuilder = new ConditionQueryBuilder<ModelClass>(table).setSeparator(",");
    }

    /**
     * Specifies a condition clause for this SET.
     * @param setClause
     * @param args
     * @return
     */
    public Set<ModelClass> conditionClause(String setClause, Object... args) {
        mConditionQueryBuilder.append(setClause, args);
        return this;
    }

    /**
     * Specifies the condition query for this SET.
     *
     * @param conditionQueryBuilder The condition query to use
     * @return This instance.
     */
    public Set<ModelClass> conditionQuery(ConditionQueryBuilder<ModelClass> conditionQueryBuilder) {
        mConditionQueryBuilder = conditionQueryBuilder;
        return this;
    }

    /**
     * Specifies a varg of conditions to append to this SET
     *
     * @param conditions The varg of conditions
     * @return This instance.
     */
    public Set<ModelClass> conditions(Condition... conditions) {
        mConditionQueryBuilder.putConditions(conditions);
        return this;
    }

    /**
     * Specifies a set of content values to append to this SET as Conditions
     * @param contentValues The set of values to append.
     * @return This instance.
     */
    public Set<ModelClass> conditionValues(ContentValues contentValues) {
        for(Map.Entry<String, Object> entry : contentValues.valueSet()) {
            String key = entry.getKey();
            mConditionQueryBuilder.putCondition(Condition.column(key).is(contentValues.get(key)));
        }
        return this;
    }

    /**
     * Begins completing the rest of this UPDATE statement.
     *
     * @param whereClause The whereclause to append
     * @param args        The argument bindings for the whereClause.
     * @return
     */
    public Where<ModelClass> where(String whereClause, Object... args) {
        return where().whereClause(whereClause, args);
    }

    /**
     * Begins completing the rest of this UPDATE statement.
     *
     * @param conditions The varg of conditions for the WHERE part
     * @return The where piece of this query.
     */
    public Where<ModelClass> where(Condition... conditions) {
        return where().andThese(conditions);
    }

    /**
     * Begins completing the rest of this UPDATE statement.
     *
     * @return The where piece of this query.
     */
    public Where<ModelClass> where() {
        return new Where<ModelClass>(this);
    }

    /**
     * Begins completing the rest of this UPDATE statement.
     *
     * @param whereConditionBuilder The where condition querybuilder to use
     * @return The where piece of this query.
     */
    public Where<ModelClass> where(ConditionQueryBuilder<ModelClass> whereConditionBuilder) {
        return where().whereQuery(whereConditionBuilder);
    }

    /**
     * Executes a SQL statement that retrieves the count of results in the DB.
     *
     * @return The number of rows this query returns
     */
    public long count() {
        return where().count();
    }

    @Override
    public String getQuery() {
        QueryBuilder queryBuilder =
                new QueryBuilder(mUpdate.getQuery())
                        .append("SET ")
                        .append(mConditionQueryBuilder.getQuery()).appendSpace();
        return queryBuilder.getQuery();
    }

    @Override
    public Class<ModelClass> getTable() {
        return mConditionQueryBuilder.getTableClass();
    }

    @Override
    public Query getQueryBuilderBase() {
        return mUpdate;
    }

    @Override
    public Cursor query() {
        FlowManager.getDatabaseForTable(mConditionQueryBuilder.getTableClass()).getWritableDatabase().execSQL(getQuery());
        return null;
    }

    @Override
    public void queryClose() {
        query();
    }
}

File: dbflow/src/main/kotlin/com/raizlabs/dbflow5/query/Set.kt
package com.raizlabs.dbflow5.query

import android.content.ContentValues
import com.raizlabs.dbflow5.addContentValues
import com.raizlabs.dbflow5.sql.Query
import com.raizlabs.dbflow5.structure.ChangeAction

/**
 * Description: Used to specify the SET part of an [com.raizlabs.dbflow5.query.Update] query.
 */
class Set<T : Any> internal constructor(
    override val queryBuilderBase: Query, table: Class<T>)
    : BaseTransformable<T>(table), WhereBase<T> {

    private val operatorGroup: OperatorGroup = OperatorGroup.nonGroupingClause().setAllCommaSeparated(true)

    override val query: String
        get() = " ${queryBuilderBase.query}SET ${operatorGroup.query} "

    override val primaryAction: ChangeAction
        get() = ChangeAction.UPDATE

    /**
     * Specifies a varg of conditions to append to this SET
     *
     * @param conditions The varg of conditions
     * @return This instance.
     */
    fun conditions(vararg conditions: SQLOperator) = apply {
        operatorGroup.andAll(*conditions)
    }

    /**
     * Specifies a varg of conditions to append to this SET
     *
     * @param condition The varg of conditions
     * @return This instance.
     */
    infix fun and(condition: SQLOperator) = apply {
        operatorGroup.and(condition)
    }

    fun conditionValues(contentValues: ContentValues) = apply {
        addContentValues(contentValues, operatorGroup)
    }

    override fun cloneSelf(): Set<T> {
        val set = Set(
            when (queryBuilderBase) {
                is Update<*> -> queryBuilderBase.cloneSelf()
                else -> queryBuilderBase
            }, table)
        set.operatorGroup.andAll(operatorGroup.conditions)
        return set
    }
}
